File: library/src/main/java/com/example/multidagger/ui/MainActivity.java
package com.example.multidagger.ui;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;

import com.example.multidagger.request.RequestClient;
import com.example.multidagger.request.RequestClientImpl;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "TestActivity";

    private final CompositeDisposable disposables = new CompositeDisposable();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        RequestClient request = new RequestClientImpl(this);
        disposables.add(request.getText()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(text -> Log.d(TAG, "Received message from service: " + text))
        );
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (!disposables.isDisposed()) {
            disposables.dispose();
        }
    }

}

File: library/src/main/kotlin/com/example/multidagger/ui/MainActivity.kt
package com.example.multidagger.ui

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.lifecycleScope
import com.example.multidagger.request.bindServiceAndWait
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

private const val TAG = "TestActivity"

class MainActivity : AppCompatActivity(), LifecycleOwner {
    private val mainCoroutine = CoroutineName("TestCoroutine")

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        runBlocking {
            lifecycleScope.launch(Dispatchers.Main + mainCoroutine) {
                val service = this@MainActivity.bindServiceAndWait()
                val text = service.text

                Log.d(TAG, "Received message from service: $text on ${currentCoroutineContext()[CoroutineName]?.name}")
            }
        }
    }
}
